metals/src/main/scala/scala/meta/internal/metals/WorkspaceFolders.scala

```diff
@@ -0,0 +1,66 @@
+package scala.meta.internal.metals
+
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.builds.BuildTools
+import scala.meta.internal.metals.MetalsEnrichments._
+
+class WorkspaceFolders(
+    initialFolders: List[Folder],
+    createService: Folder => MetalsLspService,
+    workspaceLspService: WorkspaceLspService,
+)(implicit ec: ExecutionContext) {
+
+  private val folderServices: AtomicReference[List[MetalsLspService]] = {
+    val res = createServices(initialFolders)
+    if (res.isEmpty)
+      new AtomicReference(List(createService(initialFolders.head)))
+    else new AtomicReference(res)
+  }
+
+  def getFolderServices: List[MetalsLspService] = folderServices.get()
+
+  def changeFolderServices(
+      toRemove: List[Folder],
+      toAdd: List[Folder],
+  ): Future[Unit] =
+    Future
```

Comment:
Shouldn't this be synchronized? What happens if this is invoked a couple times at the same time?

Reply:
It adds each service and removes each one by one. When doing so each add/remove is atomic. Removing/adding a folder is idempotent, since I never add the same folder twice (I remove the old one). We would have a problem if you removed a URI and aded it back in the next command but those would execute in the wrong order. I'll fix it.

Reply:
Please make sure it's fine now.